core/src/test/java/org/elasticsearch/discovery/EndlessIndexingLoopIT.java

```diff
@@ -0,0 +1,181 @@
+        SingleNodeDisruption disruptionNode2 = new BlockClusterStateProcessing(node_2, getRandom());
+        internalCluster().setDisruptionScheme(disruptionNode2);
+
+        logger.info("--> move shard from {} to {}", node_1, node_2);
+        internalCluster().client().admin().cluster().prepareReroute().add(new MoveAllocationCommand(new ShardId("test", 0), node_1, node_2)).get();
+
+        logger.info("--> wait for relocation to start");
+        beginRelocationLatchNode2.await();
+        // start to block cluster state processing for node_2 so that it will be stuck with the cluster state 1 in above description
+        disruptionNode2.startDisrupting();
+
+        logger.info("--> wait for relocation to finish");
+        endRelocationLatchNode2.await();
+        // now node_2 is still on cluster state 1 but will have have the shard moved to POST_RECOVERY
+        final Client node1Client = internalCluster().client(node_1);
+        final Client node2Client = internalCluster().client(node_2);
+        // wait until node_1 actually has removed the shard
+        assertBusy(new Runnable() {
+            @Override
+            public void run() {
+                ClusterState clusterState = node1Client.admin().cluster().prepareState().setLocal(true).get().getState();
+                // get the node id from the name. TODO: Is there a better way to do this?
+                String nodeId = null;
+                for (RoutingNode node : clusterState.getRoutingNodes()) {
+                    if (node.node().name().equals(node_1)) {
+                        nodeId = node.nodeId();
+                    }
+                }
+                assertNotNull(nodeId);
+                // check that node_1 actually has removed the shard
+                assertFalse(clusterState.routingNodes().routingNodeIter(nodeId).hasNext());
+            }
+        });
+
+        logger.info("--> cluster state on {} {}", node_1, node1Client.admin().cluster().prepareState().setLocal(true).get().getState().prettyPrint());
+        logger.info("--> cluster state on {} {}", node_2, node2Client.admin().cluster().prepareState().setLocal(true).get().getState().prettyPrint());
+        logger.info("--> index doc");
+        Future<IndexResponse> indexResponseFuture = client().prepareIndex("test", "doc").setSource("{\"text\":\"a\"}").execute();
+        // wait a little and then see how often the indexing request was sent back and forth
+        sleep(1000);
+        // stop disrupting so that node_2 can finally apply cluster state 2
+        logger.info("--> stop disrupting");
+        disruptionNode2.stopDisrupting();
+        clusterHealth = client().admin().cluster().prepareHealth()
+                .setWaitForRelocatingShards(0)
+                .get();
+        assertThat(clusterHealth.isTimedOut(), equalTo(false));
+        indexResponseFuture.get();
```

Comment:
hm. so this hangs now every now and then for a minute. I think it is when the coordinating node is node_1. Then the cluster state observer waits for the next cluster state which does not come and the index request is only executed when the observer times out.
We can send the request via node_2 but I think we actually need a way to handle this better.